### PR TITLE
[React16] Attempt to convert `React.createClass()` to ES6 classes

### PR DESCRIPTION
**Pull request by resir014**
Reference: #437.

Note that this will most likely give out CI errors in the first place, hence the [WIP] tag.

### Known issues

* ~~Some `PropTypes` checks fail (e.g. in `MusicSelectPreviewer`), therefore song previews don't currently load~~ This actually works now.

### Other notable changes

* Converted some components with `recompose/pure` mixins to `React.PureComponent`
* Changed some createClass components to SFCs
* Fixed incorrect `PropTypes`.
* Other minor tweaks + code cleanups, such as addign `key` attributes for any iterated elements.

### CODE REVIEW

**Comment by resir014** [comment]
@dtinth Everything should be playable again now! I'll take out the WIP tag and give you a chance to try out the branch and see if I missed anything.

**Comment by dtinth** [comment]
@resir014 Thanks! Don’t worry about coverage score 😄

File: src/app/ui/MusicSelectScene.jsx [hunk 1]
```diff
@@ -283,53 +288,53 @@ export const MusicSelectScene = React.createClass({
...
+  }
+  handleLogout = () => {
     if (confirm('Do you really want to log out?')) {
       Promise.resolve(online.logOut()).done()
       Analytics.send('MusicSelectScene', 'logout')
     }
-  },
-  handleAuthenticate () {
+  }
+  handleAuthenticatev = () => {
```
Comment (dtinth) [inline src/app/ui/MusicSelectScene.jsx]: I think this is misspelt — it broke the log-in/sign-up dialog.

Comment (resir014) [reply src/app/ui/MusicSelectScene.jsx]: Ah, whoops, totally missed that!

File: src/app/ui/ChangelogPanel.jsx [hunk 2]
```diff
@@ -6,17 +6,20 @@ import _        from 'lodash'
 import Panel    from 'bemuse/ui/Panel'
 import Markdown from 'bemuse/ui/Markdown'
 
-export const ChangelogPanel = React.createClass({
-  getInitialState () {
-    return { data: { status: 'loading' } }
-  },
+class ChangelogPanel extends React.Component {
+  constructor () {
+    super()
```
Comment (dtinth) [inline src/app/ui/ChangelogPanel.jsx]: Actually the constructor form should be:

```js
  constructor (props) {
    super(props)
```

Although [there is no difference if props is not accessed during construction](https://stackoverflow.com/a/34995257/559913), the documentation recommends [always passing `super(props)`](https://reactjs.org/docs/state-and-lifecycle.html#adding-local-state-to-a-class).

However, if all a constructor does is setting the state, it is better to remove a constructor entirely and use public class fields:

```js
  state = { data: { status: 'loading' } }
```

File: src/ui/YouTube.jsx [hunk 3]
```diff
@@ -3,24 +3,28 @@ import './YouTube.scss'
 import React from 'react'
 import ReactDOM from 'react-dom'
 
-export default React.createClass({
+export default class extends React.Component {
```
Comment (dtinth) [inline src/ui/YouTube.jsx]: Component is missing name? e.g. (`class YouTube extends`)

File: bin/create-component [hunk 4]
```diff
@@ -13,11 +13,11 @@ import './名.scss'
 import React from 'react'
 import c     from 'classnames'
 
-export default React.createClass({
+export default class extends React.Component{
```
Comment (dtinth) [inline bin/create-component]: Wow, didn’t thought that you will also change this script. Awesome!!

This is almost perfect, but I prefer adding a name to the class (e.g. `export default class 名 extends`)

Comment (dtinth) [inline bin/create-component]: Missing space before `{`.

**Pull request merged**